### DISCUSSION TITLE
Struct augmentation

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5897] [Language Ideas]
As discussed in #524, we allow "augmenting" structs, for the purposes of code reusability.
1. This is **not** inheritance like classes. 
2. We allow "upcasting" using an implicit operator as a convenience, but that loses the added state of the child struct; there is no "downcasting" (from base to child). 
3. C# can use the emitted attribute to make it look like "base" members are also available on the "child"; other languages just see the "base" struct as a field, which they can access, as usual. 
4. I purposefully use the word "augment", and not "inherit" (to avoid confusion with class-style inheritance, which is impossible with structs), and not "extend" (to avoid confusion with extensions, both existing extension methods and #5497 extensions).
5. Multiple "base" structs can be augmented.
```cs
struct Point2D : IEquatable<Point2D>
{
  public Point2D(double x, double y) { this.X = x; this.Y = y; }
  public double X, Y;
  public bool Equals(Point2D obj) { ... }
  public Point2D Add(Point2D obj) => new Point2D(this.X + obj.X, this.Y + obj.Y);
}
struct Point3D augments Point2D : IEquatable<Point3D>
{
  public Point3D(double x, double y, double z) : base(x, y) { this.Z = z; }
  public double Z;
  public int Point2D; // compile error - this field name is reserved
  public bool Equals(Point3D obj) { ... }
}
// which is compiled to
[AugmentedFrom("Point2D")] // the string has the field name
struct Point3D : IEquatable<Point2D>, IEquatable<Point3D> // interfaces are "lifted"
{
  public Point3D(double x, double y, double z) { this.Point2D = new Point2D(x, y); this.Z = z; }
  public Point2D Point2D;
  public double Z;
  public bool Equals(Point2D obj) => this.Point2D.Equals(obj); // have to define interface implementing methods, for boxing purposes
  public bool Equals(Point3D obj) { ... }
  public static implicit operator Point2D(Point3D obj) => obj.Point2D;
}
var p3 = new Point3D(1, 2, 3);
Point2D p2 = p3; // p2.X == 1, p2.Y == 2
p3 = (Point3D)p2; // compile error - no conversion exists (unless defined manually)
double x = p3.X; // becomes p3.Point2D.X
bool b = p3.Equals(p2); // becomes p3.Point2D.Equals(p2)  for performance - 
// no need to call p3.Equals(p2), which was created solely for the interface and boxing
p2 = p3.Add(p2); // becomes p3.Point2D.Add(p2)
object p3Boxed = p3;
var p2Equatable = (IEquatable<Point2D>)p3Boxed; // works - because we "lifted" the interfaces
var p2Unboxed = (Point2D)p3Boxed; // this can be made to work, but would require special logic similarly to Nullable<T>
p2Unboxed = object.UnboxPromoted<Point2D>(p3Boxed); // a new static helper method would be the easy solution here

// ref on "base" struct
void Foo(ref Point2D p) { ... }
Foo(ref p3); // becomes Foo(ref p3.Point2D)

// red on "base" struct's field
void Bar(ref double x) { ... }
Bar(ref p3.X); // becomes Bar(ref p3.Point2D.X)

// multiple structs
struct Point { ... }
struct Size { ... }
struct Rectangle augments Point, Size {
  public Rectangle(int x, int y, int width, int height) : Point(x, y), Size(width, height) { }
}

// extensions/roles
extension Point2DExt : Point2D { void Foo() { ... } }
p3.Foo(); // becomes p3.Point2D.Foo() - or however that's implemented, per roles/extensions feature
public static class Point2DExtOld { public static void Bar(this Point2D This) { ... } }
p3.Bar(); // becomes Point2DExtOld.Bar(p3.Point2D)

abstract struct WrapperBase<T>
{
  public WrapperBase(T value) { this._value = value; }
  T _value;
  public T Value => this._value;
}
struct Weight augments WrapperBase<double>
{
  // no ctor needed - base ctor is propagated automatically
  public Density GetDensity(Volume volume) => new Density(this._value / volume.Value); // custom methods
}
struct Length augments WrapperBase<double>
{
  public Length(double value, LengthUnit lengthUnit) : base(value) { this.LengthUnit = lengthUnit; }
  public LengthUnit LengthUnit { get; } // added domain-specific state
}
var w = new Weight(1.2);
var l = new Length(2.3, LengthUnit.Meter);

// another useful scenario - currently I create all these as classes, for inheritance, but these are unnecessary allocations
struct CreateOrderRequest
{
  string product;
}
struct CreateRecurringOrderRequest augments CreateOrderRequest
{
  int numDaysToRepeat;
}
void CreateOrder(CreateOrderRequest request) { ... }
void CreateRecurringOrder(CreateRecurringOrderRequest  request) { ... }
```
Open questions:
1. If we add multi-struct augmentation, do we retain implicit operators to each of the "base" structs? It can be weird to convert a `Rectangle` to `Point` in the above example. Should it be an `explicit` operator maybe, or no operator at all?
2. Do we allow the creation of `abstract struct`s? If so, what are the rules? The only possible limitation can be inability to call the constructors, but `default` can always be assigned to a local variable, for example. Or a blanket ban on using anywhere, except for "augmentation"?
3. Is it worthwhile to introduce a new keyword, `augments`, or just good old `:` will do? The upside of `augments` is that it's clear that this is not class-style inheritance; the downside is it's verbose, and it breaks from C# style - becomes more Java-like, which isn't necessarily a bad thing, but it's a change nonetheless.

If anybody has questions, I will try to answer and update this post, as needed.

**Comment by Thaina** [comment]
Can this be pass by ref ?

Suppose there is

```C#
void DoSomething(ref Point2D point)
{
    point.X = SomeValX;
    point.Y = SomeValY;
}

Point3D p;
DoSomething(ref p);
```

Possible?

**Reply by TahirAhmadov** [reply]
No, according to current C# syntax. Also, we can't even make it work if we wanted to: the implicit operator returns a value, and we cannot pass in a ref to a return value.
On second thought, maybe we can add logic to the compiler to sidestep the implicit operator, and just do a ref to the Point2D field directly? @CyrusNajmabadi , help :)

**Reply by Thaina** [reply]
@TahirAhmadov One possibility is to transpile `Augmented` when it pass by ref

```C#
Point3D p;
DoSomething(ref p);

//// become

DoSomething(ref p.Point2D); // taking the field that it augmented from

//// at compile time
```

**Reply by TahirAhmadov** [reply]
I updated the OP to cover this, nobody said this cannot be done, so for now we assume it can be. :)

**Comment by HaloFour** [comment]
Feels like something that would be very easily implemented with a source generator:

```cs
struct Point2D : IEquatable<Point2D>
{
  public Point2D(double x, double y) { this.X = x; this.Y = y; }
  public double X, Y;
  public bool Equals(Point2D obj) { ... }
  public Point2D Add(Point2D obj) => new Point2D(this.X + obj.X, this.Y + obj.Y);
}

[Augments<Point2D>]
partial struct Point3D : IEquatable<Point3D>
{
  public Point3D(double x, double y, double z) {
    this.Point2D = new(x, y);
    this.Z = z;
  }
  public double Z;
  public bool Equals(Point3D obj) { ... }
}

// source generator emits
partial struct Point3D : IEquatable<Point2D>
{
  public Point2D Point2D;
  public bool Equals(Point2D obj) => this.Point2D.Equals(obj); // have to define interface implementing methods, for boxing purposes
  public static implicit operator Point2D(Point3D obj) => obj.Point2D;
}
```

**Reply by TahirAhmadov** [reply]
That won't compile? You can't call `: base(x, y)` like that.

**Reply by HaloFour** [reply]
Copypasta on my part.  Yes, it couldn't have a call to a `base` ctor.

**Reply by TahirAhmadov** [reply]
That's actually pretty close to how it would look "natively". I guess the only 2 downsides to source generator approach is 1) lack of standardization - language feature makes it available to everybody, and 2) the minuscule stylistic imperfection of not having `: base(x, y)`.

**Reply by TahirAhmadov** [reply]
Another downside - see @Thaina 's question above - if we can make `ref` work through trans-piling, 3) source generator cannot give you that.

**Reply by TahirAhmadov** [reply]
Even after your update to set `Point2D` field explicitly, will this compile? The field isn't declared in the code itself - can you reference members which **will** become available after source generators are run?

**Reply by RikkiGibson** [reply]
Yes. SGs expect to run in presence of errors such as a member "which will be generated" not being present.

**Reply by TahirAhmadov** [reply]
So I guess the downside here is lack of editor features? Auto-complete, etc.

**Reply by RikkiGibson** [reply]
The editor is using the "after SGs are run" view of the compilation. The downside of this is that the performance of the editor can be negatively impacted by misbehaving/poorly optimized source generators.

**Comment by CyrusNajmabadi** [comment]
I'm curious if we can go a *much* simpler route.  Specifically, have the language just *automatically* lift members of nested-structs upwards as long as they do not collide with real members in the name-spaces they are lifted to.  So, in other words, instead of the above, you just write:

```c#
struct Point2D : IEquatable<Point2D>
{
  public Point2D(double x, double y) { this.X = x; this.Y = y; }
  public double X, Y;
  public bool Equals(Point2D obj) { ... }
  public Point2D Add(Point2D obj) => new Point2D(this.X + obj.X, this.Y + obj.Y);
}
struct Point3D : IEquatable<Point3D>
{
   public Point2D Point2d;

  public Point3D(double x, double y, double z) { Point2d = new(x, y); this.Z = z; }
  public double Z;
  public bool Equals(Point3D obj) { ... }
}
```

And Point3d automatically exposes `X`, `Y`, `Add` etc. from the nested struct.  

Put another way, name lookup on Point3d would work by looking up in Point3d first, then inside the nested structs within.

If doing this automatically was undesirable, it could be done through a keyword like `promote`.  So: 

```c#
struct Point3D : IEquatable<Point3D>
{
   promote public Point2D Point2d;
```

**Reply by TahirAhmadov** [reply]
The difference with your approach (either automatic or with `promote`) is that we then allow multiple "base" structs. I'm not saying it's a bad idea, but definitely something to think through. At least, it's inconsistent with classes where only one class can be inherited (yes, I know, this isn't inheritance, but still.)

**Reply by CyrusNajmabadi** [reply]
> is that we then allow multiple "base" structs. I'm not saying it's a bad idea, but definitely something to think through.

I like that idea, and that's particularly why i'm not treating this as anything like 'inheritance'.  It's just promoted name lookup to make things more pleasant. :)

**Reply by TahirAhmadov** [reply]
Will `promote` also bring interfaces up to the "child"? What if there are conflicts between regular members (fields, methods, etc.)? What if there are conflicts between interface implementations?

**Reply by CyrusNajmabadi** [reply]
I view it just as name lookup. For things like lifting the interface I would prefer a different feature entirely (e.g. implement an interface through a member). 

> What if there are conflicts between regular members (fields, methods, etc.)? 

The same as any name lookup.  :-)
This is why I'm somewhat amenable to declaring explicitly which subfields to promote. 

> What if there are conflicts between interface implementations?

Does not apply afaict.

**Reply by TahirAhmadov** [reply]
I'm not sure I understand what you mean by "name lookup". Say, 2 structs are promoted, both of which have a member with the same name. Is that allowed? Which one is used? The first one, as declare in code? If so, what happens with partial structs, one promoted struct is one partial, and another promotion in another partial?
What if one of the promoted structs did not have a member, but a new version of the DLL does? Do we have to add logic to the binder to detect this collision?
And removing interfaces kind of takes away from the feature.
I don't know, to be honest. Sounds like we're introducing some of the difficulties of multiple inheritance, while at the same time taking away one of the core advantages - interfaces.

**Reply by CyrusNajmabadi** [reply]
>I'm not sure I understand what you mean by "name lookup". Say, 2 structs are promoted, both of which have a member with the same name. Is that allowed?

If it's a basic name, it would cause a collision needing disambiguation.  If it's a signature, it would only be an issue if overload-resolution could not determine a best match.

> If so, what happens with partial structs, one promoted struct is one partial, and another promotion in another partial?

Partial would have no effect on anything.  That's just a mechanism to split up the declaration of a symbol.  it would not affect the name lookup of members inside of it.

> What if one of the promoted structs did not have a member, but a new version of the DLL does? Do we have to add logic to the binder to detect this collision?

All binding happens statically at compile time.  

This is the same thing that happens today with any symobls.  Consider the following:

```c#
// assembly 1

class A
{
     public void M() { }
}

class B : A
{
}

// assembly 2
new B().M();
```

This calls into A.M statically .  If a new version of assembly1 updates to be:

```c#
class B : A
{
    public new void M() { }
}
```

tehn nothing changes in assembly2.  All the resolution and binding happened at compile time.  Changing the shape without recompiling downstream callers won't lead to any changes downstream.

> And removing interfaces kind of takes away from the feature.

I don't really see why yet :)   Your own example doesn't really even show why that's important.

**Reply by TahirAhmadov** [reply]
Regarding collisions - OK I guess we can make all that work, but is it worth the complication? We would have to define new error(s), test all these scenarios, etc. - and for what? Is there a use case to "promote" multiple structs?
While "augment" is not the same as "inherit", it follows a similar style which is known by everybody, whereas the "promote" syntax is completely new.
Regarding interfaces - it's arguably the most important part of this. With the `WrapperBase<T>` and `Length` etc. structs, I see `WrapperBase<T>` implementing a few interfaces - reusing that implementation is the crux of the matter. Also, with static interface members, you could have numeric operations which are implemented once and available to all the physical measurement types.
Also, I suppose it would have to be prohibited to "promote" more than 1 field of the same type?
One more downside of "promote" - the constructor has to explicitly initialize the field(s), instead of the usual `:base(...)` syntax.
And the fields being "visible" (meaning, you can do `p3.X` or `p3.Point2D.X`) is more confusion. If I were to see that surface area, I would immediately ask - are those different things, or the same thing.
Still feels very complicated.

**Reply by CyrusNajmabadi** [reply]
> and for what? Is there a use case to "promote" multiple structs?

Sure.  You're building a compositive out of multiple smaller pieces :)

> Regarding collisions - OK I guess we can make all that work, but is it worth the complication?

From my perspective it's basically equally complex.  You look up in the normal type first, then in all the promoted types (which then recursively falls out for further levels of nesting).

> Also, I suppose it would have to be prohibited to "promote" more than 1 field of the same type?

Yeah, since it would almost always not be helpful, that seems reasonable to me.

**Reply by TahirAhmadov** [reply]
If you think multiple "base" structs are valuable, why not add it to the "augment" syntax?
```cs
struct Child augments Base1, Base2
{
  public Child(int x, int y) : Base1(x), Base2(y) {  }
}
```
This is win-win-win: we get the "base ctor" invocation; we get the interfaces; and we don't show the actual fields (at least in C# we don't), preventing confusion.

**Reply by TahirAhmadov** [reply]
Also, with "augment", the compiler can apply extensions and roles more "naturally". With "promoted fields", it _can_ be done, but looks awkward.
```cs
struct Point2D { }
struct Point3D augments Point2D { }
extension Point2DExt : Point2D
{
  void Foo() { ... }
}
var p3 = new Point3D();
p3.Foo(); // becomes p3.Point2D.Foo()
```

**Reply by RikkiGibson** [reply]
It seems like there's no need to restrict the containing type to being a struct either, i.e. perhaps it would be reasonable to let such "promoted"/"flattened" fields be declared within class types as well.

**Reply by TahirAhmadov** [reply]
@RikkiGibson that sounds like a completely different feature to what I envisioned. If we allow "promoting" fields for classes, then we can't really do any of the struct-specific improvements which are valuable. To me, just "flattening" the fields is nowhere near worth doing anything at all.

**Reply by RikkiGibson** [reply]
> If we allow "promoting" fields for classes, then we can't really do any of the struct-specific improvements which are valuable.

It wasn't clear to me why this is the case, could you elaborate?

**Reply by TahirAhmadov** [reply]
If you look at my proposal, it includes things like interface lifting, one-way "upcast" using an implicit operator, passing "base" struct as `ref` (thanks to Thaina), etc. None of these things really make sense with classes. For interfaces, it can be done, but it goes against the existing inheritance hierarchy. For "upcasting", it's normal to expect a class to be "castable" to either it's base or child classes (if instance is there), and implemented interfaces; now we're introducing a whole new way of "casting" which is different from how classes work. Same goes for `ref` - if you have an instance of a class, which has a class-type field, it would be completely unexpected that that field is passed in by `ref` instead of the class itself.
These reasons are also why I'm unsure about the propriety of using "promoted fields" even for structs; I think of it as "struct augmentation", because it's similar (but not the same) as inheritance, for the struct world. Also for things like extensions, etc. - these don't make sense with promoted fields.

**Comment by CyrusNajmabadi** [comment]
I think it would be helpful to not only say what would be possible here but also the use cases why it would be valuable.  For example the interface/ref stuff just doesn't seem that major.  How often do you need to do this, and in those cases, would it be that bad to just say `OuterType.InnerProp` instead of just passing OuterType naked?

**Reply by TahirAhmadov** [reply]
Regarding interfaces - it's arguably the most important part of this. With the `WrapperBase<T>` and `Length` etc. structs, I see `WrapperBase<T>` implementing a few interfaces, like `IEquatable<>`, etc. - reusing that implementation is the crux of the matter. Also, with static interface members, you could have numeric operations which are implemented once and available to all the physical measurement types.
For `ref`, the importance stems from the ability to use `ref` with structs to modify them in place. If we are to agree that passing a struct by `ref` is important, it follows that being able to easily re-use a method made for a "base" struct, to modify a "child" struct (specifically, it's "base" field), is also important. I view this as being very similar to being able to pass in a class to a method accepting its base class - surely that's very helpful :) In both cases, you have an object, inherited for classes and augmented for structs, and you can alter it by passing it to a method which knows how to deal with the base class or "base struct".

**Reply by CyrusNajmabadi** [reply]
@TahirAhmadov all of these thigns seem possible by just passing `Outer.Inner`.  where the breakdown for me is showing enough necessity that this works just passing `Outer`.

**Reply by TahirAhmadov** [reply]
You can't do that with interfaces. It doesn't work when boxed without auto-lifting the interfaces.
For `ref`, it's not the end of the world if we don't have it, but I think it's an easy thing to add, and it also has the value of making the feature "complete". 
Again, we're not aiming for class-style inheritance, but this augmentation approach - with the interfaces, the ref, the extensions, the implicit operator, - gives us many similar benefits, while being safe and easy to use.

**Reply by CyrusNajmabadi** [reply]
> You can't do that with interfaces. It doesn't work when boxed without auto-lifting the interfaces.

I'm not sure what you mean here.  

> with the interfaces, the ref, the extensions, the implicit operator, - gives us many similar benefits, while being safe and easy to use.

I guess i'm asking to speak with less vagueries :)  What in particular do you get?  Can you whip up an example?

**Reply by TahirAhmadov** [reply]
For `ref`, I see a typical use case like this:
```cs
void MultiplyDependingOnCondition(ref LargeStruct<double> value)
{
  if(condition)
  {
    value.Num1 *= coefficient;
    value.Num2 *= coefficient;
    value.Num3 *= coefficient;
    value.Num4 *= coefficient;
  }
}
```

**Reply by TahirAhmadov** [reply]
Please look through the OP, I think I covered everything in it.

**Reply by TahirAhmadov** [reply]
I added a code snippet in the middle which explains what I mean by interfaces:
```cs
var p2Equatable = (IEquatable<Point2D>)p3Boxed; // works - because we "lifted" the interfaces
```

**Comment by Xyncgas** [comment]
but, just inherit the struct man

**Reply by Xyncgas** [reply]
oh you can't?

**Reply by Xyncgas** [reply]
We should be able to inherit a struct, have virtuals in the struct, override them or make them invisible like we can do with classes (new, override...)

For example
```
public struct StructA
{
    [ExplicitLayout(0)]
    public int MyInt32_0;
    [ExplicitLayout(4)]
    public int MyInt32_1;
}
public struct StructA2:StructA
{
    public Masked(0,8) Int64 MyInt64; //now MyInt32_0 MyInt32_1 turns invisible
}
```
meh maybe it's a horrible idea

**Reply by CyrusNajmabadi** [reply]
> have virtuals in the struct,

structs have no vtable.  there's no place to put that.  That's the nature of a struct, it's just the storage for its inside data.  If you want inheritance/virtuals/etc., that's already solved with classes.

**Comment by Xyncgas** [comment]
But in general what if you just use explicit layout, for example if I want to augment point2D in point3D just declare ```x,y,z``` in point3D like you normally would but also declare Point2D inside the Point3D struct and have it overlap with ```x,y``` so now you can use Point2D to access only the two values

**Reply by Xyncgas** [reply]
in fact you can even augment Vector3 into your point3D and basically gives your Point3D struct SIMD capabilities

**Reply by Xyncgas** [reply]
just overlap your fields, like in c++, or just forget about all this and take a reference and cast to another pointer type and now you have access to any member functions you want since you can cast your void* to any pointer

**Comment by TahirAhmadov** [comment]
@CyrusNajmabadi I just ran across a use case for "promoted" fields (or even properties) the way that you envisioned them, which is separate from my struct augmentation idea. Specifically, imagine your typical business software with data entities; it's frequently necessary to reuse certain properties between multiple entities, but lack of multiple inheritance prevents us from doing so elegantly. I think "promotion" would be the next best thing:
```cs
interface IOrganization
{
  promoted LegalEntityDescriptor Descriptor { get; set; }
}
class Company : IOrganization
{
  public string Name { get; set; } // if this is present, then the line below generates a compile error
  public promoted LegalEntityDescriptor Descriptor { get; set; }
}
class Charity : IOrganization
{
  public promoted LegalEntityDescriptor Descriptor { get; set; }
}
class LegalEntityDescriptor // or struct
{
  public string Name { get; set; }
  public string MailingAddress { get; set; }
}
```
I think it makes sense to go with explicit approach here: 
- at compile time, conflicts are prohibited; 
- accessing the members of the promoted type through the field/property is a warning (but getting/setting said field/property is allowed);
- promoted properties on interfaces are allowed, too, and implementing class has to follow the "promoted" rule on the property.
- if promoted field/property is nullable, then this won't work; meaning, it has to be non-null, if class, or it has to be struct.

Thoughts?

**Reply by HaloFour** [reply]
Sounds very easily accomplishable via source generators.

**Reply by TahirAhmadov** [reply]
> Sounds very easily accomplishable via source generators.

True, perhaps even more so, than my struct augmentation idea. A combination of a source generator and an analyzer should be able to make this work.

**Reply by TahirAhmadov** [reply]
Can source generators change manually entered code? If I write, `company.Name`, will it be virtually changed to `company.Descriptor.Name`? Or were you thinking that a pass-through property `Name` property would be defined on `Company` class, etc., which would access `this.Descriptor.Name`?
The problem with auto-creating said pass-through properties is that in reflection scenarios, this can create a problem when certain things are done for each property of a class; especially if it's third party libraries, it may be difficult to exclude these.
On the other hand, the presence of these auto-generated pass-throughs may be beneficial elsewhere.

**Reply by HaloFour** [reply]
The source generator would generate delegating properties:

```cs
partial class Company : IOrganization
{
  [Promoted] public LegalEntityDescriptor Descriptor { get; set; }
}

// generated
partial class Company {
    public string Name => this.Descriptor.Name;
    // etc.
}
```

**Reply by HaloFour** [reply]
> The problem with auto-creating said pass-through properties is that in reflection scenarios

I don't see why this is a problem.  Delegation is a pretty common design pattern.  This is exactly what I would expect any "first class" compiler support for such a thing to do.

**Reply by TahirAhmadov** [reply]
> I don't see why this is a problem. Delegation is a pretty common design pattern. This is exactly what I would expect any "first class" compiler support for such a thing to do.

I actually think it would be the opposite for native compiler support; it makes no sense to add stuff to metadata which isn't in the code. And that, in my mind, creates a problem - if I was to write such an SG, but the language would later get the feature, it would likely be incompatible in some ways.

**Reply by HaloFour** [reply]
> I actually think it would be the opposite for native compiler support

I disagree.  Delegation is a very common pattern and exactly how I would expect any such feature to be implemented.  I would want those members to be directly discoverable, and I would want them to be able to participate in such things as interface implementation automatically.  It also puts a lot less onus on any consuming language that now doesn't need to know anything about the "magic" of promoted members, they're just normal members like any other.

> And that, in my mind, creates a problem - if I was to write such an SG, but the language would later get the feature, it would likely be incompatible in some ways.

I seriously doubt that delegation will ever become a language feature.  It's so easily solvable by tooling and source generators that I don't see why it would bubble up to the language level.

**Reply by TahirAhmadov** [reply]
> I disagree. Delegation is a very common pattern and exactly how I would expect any such feature to be implemented. I would want those members to be directly discoverable, and I would want them to be able to participate in such things as interface implementation automatically. It also puts a lot less onus on any consuming language that now doesn't need to know anything about the "magic" of promoted members, they're just normal members like any other.

What you are saying makes sense, but I was under the impression that they don't want to create stuff in IL that doesn't exist in code. Perhaps this delegation approach does not fall under that umbrella; if so, the benefits you outlined are worthwhile, indeed. 
Also, please see my idea below - thoughts?

**Comment by TahirAhmadov** [comment]
On second thought, I'm actually now thinking both the "augmentation" and the "promotion" can be combined into one, sweeping improvement.
```cs
class Person : EntityBase, IEquatable<Person>, promotes LegalEntityDescriptor
{
  // ...
}
struct Point3D : IEquatable<Point3D>, promotes Point2D // for structs, this is the same as augments above
{
  // ...
}
```
In fact, classes can promote structs and vice versa, as needed. This solves (for most use cases) two very important and fundamental shortcomings of C# today: lack of multiple inheritance and lack of struct inheritance. It allows greater use of structs to reduce allocations, where desirable.

**Reply by TahirAhmadov** [reply]
Before writing an SG to imitate this, I tried just writing the code by hand, and here are the limitations of that approach:
- Syntax awkwardness:
  - no ` : base(x, y)` constructor calls (more likely, with language support it would be ` : Point2D(x, y)` because multiple types can be promoted);
  - the need to make the type partial;
- Impossible to promote fields
  - wrapping them in a delegation property works in some cases and not in others, for example, cannot pass by `ref`;
- None of the extension scenarios work;
- Passing "child" type to a method which expects `ref` parameter of "base" (or promoted) type doesn't work;
- Difficulties with `protected` members, without language support, these have to be exposed somehow, most likely with a special shim class, further complicating the solution and adding non-optimizable indirections;
- Similarly, `virtual`/`abstract` members are difficult to implement, with complicated shimming and awkward syntax.

In short, I think a language feature is called for here, and an SG will not suffice.

**Reply by HaloFour** [reply]
> In short, I think a language feature is called for here, and an SG will not suffice.

IMO, this is still a solution in search of a problem.  Yes, there is friction using other tools or patterns to accomplish similar goals, but those goals are not common.  If there was a desire to rework structs in C# in a meaningful way as you describe I'd much rather that be considered at the runtime level instead of the compiler level anyway as even at the compiler level there will be observable warts.

**Reply by TahirAhmadov** [reply]
> IMO, this is still a solution in search of a problem. 

I think you yourself created a discussion at some point asking for `WrapperBase<T>` type of struct for physical values, etc. I think the problem is real - the OP lists three scenarios that I thought of, and there are potentially others.

> Yes, there is friction using other tools or patterns to accomplish similar goals, but those goals are not common. If there was a desire to rework structs in C# in a meaningful way as you describe I'd much rather that be considered at the runtime level instead of the compiler level anyway 

How could it be solved at runtime level? Short of moving in the direction of C++ where classes can be value types, effectively removing the distinction (which is not happening), I don't see what else can be done.

> as even at the compiler level there will be observable warts.

I wouldn't call them warts, rather - different expectation and realities, caused by the fundamental differences between classes and structs. Also, if we say classes can be promoted, too, it's no longer just a struct improvement.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> I think you yourself created a discussion at some point asking for `WrapperBase<T>` type of struct for physical values, etc. I think the problem is real - the OP lists three scenarios that I thought of, and there are potentially others.

As a type-safe wrapper of a single value, sure, but it never attempts to pretend that the wrapper is extends from that value type.  I also created it before `struct record` was a thing, so it's pretty much obsolete now.

> I wouldn't call them warts, rather - different expectation and realities, caused by the fundamental differences between classes and structs. Also, if we say classes can be promoted, too, it's no longer just a struct improvement.

Warts still the same, and still nothing that can't be solved via delegation.  Your points about it not working with `ref` are, IMO, intentional and important.  Even without this feature odds are those would be properties and not fields anyway, precluding the ability to take a `ref`.

IMO a lot of these kinds of proposals come from an idea that references and allocations are always bad and should be avoided at all costs, so let's put a lot of effort into trying to make structs into classes instead, not realizing that passing around these big structs would end up being more expensive than those allocations.

**Reply by TahirAhmadov** [reply]
> As a type-safe wrapper of a single value, sure, but it never attempts to pretend that the wrapper is extends from that value type. I also created it before `struct record` was a thing, so it's pretty much obsolete now.

If you look at the OP, you can see how it's beneficial to actually have the semantics of "extending" (or "augmenting" or "promoting") as opposed to just a simple wrapper. For example, you can have structs with physical units and multiplying certain units produces a certain other unit, meaning, multiplying watt by second produces a joule, and not a calorie, etc.

> Warts still the same, and still nothing that can't be solved via delegation. Your points about it not working with `ref` are, IMO, intentional and important. Even without this feature odds are those would be properties and not fields anyway, precluding the ability to take a `ref`.

That may or may not be true in all use cases. I can see how math-heavy workloads want to write simple structs with fields and don't want to bother with interfaces and properties and all that. Also, the other `ref` scenario - passing the entire "base" ("promoted") struct by-ref - is also impossible in a clean manner without language support.

> IMO a lot of these kinds of proposals come from an idea that references and allocations are always bad and should be avoided at all costs, so let's put a lot of effort into trying to make structs into classes instead, not realizing that passing around these big structs would end up being more expensive than those allocations.

I agree that allocations are not automatically all bad, however the way I see "promotion" working (as opposed to the original "augmentation"), it's usable for both classes and structs. Also, for structs, it would be the users' job to figure out when to use structs, as opposed to classes; the language does not mandate either way - we're just making it easier by making the struct option more complete.
For classes, it would give us a decent approximation of multiple inheritance, without any of the downsides of actual multiple inheritance. I regularly (not SUPER frequently, like daily, more like once or twice a month) have to deal with repeated code due to lack of multiple inheritance (or LoMI, as I called a special attribute in my projects). Currently, the best thing I could think of is create ugly static classes with some "common" methods, together with interfaces. All of that is terrible. (The other alternative of copying and pasting code in multiple places is much worse than terrible.)

**Reply by TahirAhmadov** [reply]
@HaloFour I just found this item you opened: https://github.com/dotnet/csharplang/discussions/1014
What can I say, great minds think alike :)
Basically, what I think C# needs is "mixins". Your idea was a special new construct called "mixin"; my thinking was, any class/struct can be "mixed in" (promoted) into any class/struct.
I thought about it a little more and realized there is another potential here:
```cs
class Control
{
  public Point Location { get; set; }
}
struct TextControlCommon
{
  public string Text { get; set; }
  public void Draw(Graphics g) 
  {
    g.DrawString(this.Location, this.Text); // oops, Location is a property of Control 
  }
}
class Button : Control, promotes TextControlCommon
{
  public override void Draw(Graphics g)
  {
    this.TextControlCommon.Draw(g); 
  }
}
```
How do we make `Location` available to `TextControlCommon`? One option is:
```cs
struct TextControlCommon promotedby Control, ISomeInterface { ... }
// now all members of Control and ISomeInterfaces are available
// and TextControlCommon is no longer usable directly
```

**Comment by TahirAhmadov** [comment]
To anybody else who is interested in this: I did end up writing a source generator to imitate this behavior, and while it has a few downsides compared to a native language solution to allow "promotion"/"augmentation"/"mixins" - I decided to name it "composition" - it does make quite a few scenarios a lot easier.

[SourceGenerator.cs](https://github.com/dotnet/csharplang/files/13296147/SourceGenerator.cs.txt)

<details>

<summary>Supporting types</summary>

```cs
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct, AllowMultiple = true)]
public class ComposedOfAttribute : Attribute
{
	public ComposedOfAttribute(Type type)
	{
		this.Type = type;
	}

	public Type Type { get; }
}

[AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct)]
public class CompositionPartAttribute : Attribute
{
}

public interface IComposedOf<T>
{
	T Base { get; }
}
```

</details>


(BTW why aren't .cs files allowed to be attached??!!...)

PS. This is the first ever analyzer/generator I wrote - any constructive criticism is very welcome.

**Reply by TahirAhmadov** [reply]
The usage is like this:
```cs
[CompositionPart]
struct Point2D : IEquatable<Point2D>
{
  public Point2D(double x, double y) { this.X = x; this.Y = y; }
  public double X, Y;
  public bool Equals(Point2D obj) { ... }
  public Point2D Add(Point2D obj) => new Point2D(this.X + obj.X, this.Y + obj.Y);
}
[ComposedOf(typeof(Point2D))]
struct Point3D : IEquatable<Point3D>
{
  public Point3D(double x, double y, double z) { this.BasePoint2D(x, y); this.Z = z; }
  public double Z;
  public bool Equals(Point3D obj) { ... }
}
var p3 = new Point3D(1, 2, 3);
Console.Write(p3.X + ", " + p3.Y + ", " + p3.Z);
```

**Comment by TahirAhmadov** [comment]
I've vastly improved the source generator and created a Github repo and Nugets for it:

https://github.com/TahirAhmadov/Composition
https://www.nuget.org/packages/TA.Util.CompositionTypes/1.0.1
https://www.nuget.org/packages/TA.Util.Composition/1.0.0

**Reply by 333fred** [reply]
Hey Tahir, I would recommend coming and talking to us in the `#roslyn` channel on https://discord.gg/csharp. Unfortunately, `ISourceGenerator` is effectively deprecated at this point, and we do not recommend using it at all. We can talk about source generator best practices there.

**Reply by TahirAhmadov** [reply]
New package Nuget is [https://www.nuget.org/packages/TA.SourceGenerators.Composition/](https://www.nuget.org/packages/TA.SourceGenerators.Composition/)

**Reply by Aniobodo** [reply]
@TahirAhmadov the nugget package generates a class-like-interface with properties backed up by fields. Thinking about boxing on the generated structs, is that really the behaviour you want?

**Reply by TahirAhmadov** [reply]
@Aniobodo the interface is meant for unboxing an already boxed struct; you don't need to use it if you have the struct itself. Please refer to the code samples.

**Reply by Aniobodo** [reply]
@TahirAhmadov  Got it. Support for readonly struct?

**Reply by TahirAhmadov** [reply]
Contributions are welcome :)

**Reply by TahirAhmadov** [reply]
BTW you may want to try it, I don't remember if I tested it, but I think it's actually supported.
